src/pmdas/bpf/modules/exitsnoop.c

```diff
@@ -0,0 +1,468 @@
+    int err;
+    char *val;
+    int idx, cg_map_fd;
+    int cgfd = -1;
+
+    if ((val = pmIniFileLookup(cfg, module_name, "process_count")))
+        env.process_count = atoi(val);
+
+    err = ensure_core_btf(&open_opts);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to fetch necessary BTF for CO-RE: %s", strerror(-err));
+        return 1;
+    }
+
+    obj = exitsnoop_bpf__open_opts(&open_opts);
+    if (!obj) {
+        pmNotifyErr(LOG_ERR, "failed to open BPF object");
+        return 1;
+    }
+
+    obj->rodata->target_pid = target_pid;
+    obj->rodata->trace_failed_only = trace_failed_only;
+    obj->rodata->trace_by_process = trace_by_process;
+    obj->rodata->filter_cg = env.cg;
+
+    err = exitsnoop_bpf__load(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to load BPF object: %d", err);
+        return err != 0;
+    }
+
+    /* update cgroup path fd to map */
+    if (env.cg) {
+        idx = 0;
+        cg_map_fd = bpf_map__fd(obj->maps.cgroup_map);
+        cgfd = open(env.cgroupspath, O_RDONLY);
+        if (cgfd < 0) {
+            pmNotifyErr(LOG_ERR, "Failed opening Cgroup path: %s", env.cgroupspath);
+            return err != 0;
+        }
+        if (bpf_map_update_elem(cg_map_fd, &idx, &cgfd, BPF_ANY)) {
+            pmNotifyErr(LOG_ERR, "Failed adding target cgroup to map");
+            return err != 0;
+        }
+    }
+
+    err = exitsnoop_bpf__attach(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to attach BPF programs: %d", err);
+        return err != 0;
+    }
+
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            handle_event, handle_lost_events, NULL, NULL);
+    if (!pb) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to open perf buffer: %d", err);
+        return err != 0;
+    }
+
+    /* internal/external instance ids */
+    fill_instids(env.process_count, &exitsnoop_instances);
+    fill_instids(env.process_count, &exitsnoop_lost);
+
+    /* Initialize the tail queue. */
+    TAILQ_INIT(&head);
+
+    return err != 0;
+}
+
+static void exitsnoop_shutdown()
+{
+    struct tailq_entry *itemp;
+
+    free(exitsnoop_instances);
+    free(exitsnoop_lost);
+    perf_buffer__free(pb);
+    exitsnoop_bpf__destroy(obj);
+
+    /* Free the entire cache queue. */
+    while ((itemp = TAILQ_FIRST(&head))) {
+        TAILQ_REMOVE(&head, itemp, entries);
+        free(itemp);
+    }
+}
+
+static void exitsnoop_refresh(unsigned int item)
+{
+    perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
+}
+
+static int exitsnoop_fetch_to_atom(unsigned int item, unsigned int inst, pmAtomValue *atom)
+{
+    struct tailq_entry *value;
+
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    if(!get_item(inst, &value))
+        return PMDA_FETCH_NOVALUES;
+
+    /* bpf.exitsnoop.age */
+    if (item == AGE) {
+        atom->d = (value->event.exit_code - value->event.start_time) / 1e9;
```

Comment:
The calculation should include exit_**time** and not exit_**code**.
PCP includes support for different metric units and resolutions - let's use the raw resolution (nanoseconds, according to the bpf sources) and let the PCP client tools (pmval, pmrep, htop etc.) decide in which unit they want to display that metric value.
tl;dr don't divide by `1e9`, let's use the raw resolution